На ветке lesson5

RamzanBersanov opened the pull request:
Изменения, которые будут включены в коммит: задание из урока 5

RamzanBersanov committed:
На ветке lesson5
 Изменения, которые будут включены в коммит: задание из урока 5

VidgarVii commented on app/models/category.rb:
```diff
@@ -1,2 +1,3 @@
 class Category < ApplicationRecord
+  has_many :tests
```
если категорию удалить, то в текущем варианте можно в дальнейшем получить ошибку

к примеру при попытке обратиться по ассоциации (`@test.category`)
или при каких-то операциях, которые подразумевают валидацию объекта

после удаления категории в таблице `tests` останется в поле `category_id` ссылка на несуществующий объект
т.е. объекты в таблице `tests` станут невалидными: рельсы по умолчанию проверяют наличие ассоциированного объекта для всех `belongs_to` ассоциаций (при валидации этого объекта) и выбрасывают ошибку валидации, если объекта не существует

если же на уровне бд поле `category_id` обозначено, как внешний ключ, то уже на попытке удаления объекта из таблицы `categories` получим ошибку базы данных, такого примерно вида:
```
ActiveRecord::InvalidForeignKey (PG::ForeignKeyViolation: EROR:  update or delete on table "categories" violates foreign key constraint
```
и рельсы откажутся удалять категорию

поэтому для всех `has_*` ассоциаций следует указывать поведение при удалении - можно посмотреть в документации - опция `dependent: ...`

VidgarVii commented on app/models/question.rb:
```diff
@@ -1,2 +1,4 @@
 class Question < ApplicationRecord
+  has_many :answers
+  belongs_to :test
```
belongs_to и has_many разделяют пустыми строками. Такое разделение кода на логические блоки.
Пример оформления модели можно посмотреть тут https://github.com/rubocop/rails-style-guide#macro-style-methods

VidgarVii replied:
Еще одно нарушение style guide
Надо соблюдай очередность.
Принято сначала писать belongs_to далее ниже has_
Так мы будем понимать какой код где находится

VidgarVii commented on app/models/test.rb:
```diff
@@ -1,4 +1,9 @@
 class Test < ApplicationRecord
+  belongs_to :author, class_name: 'User'
+  has_many :tests_users
+  has_many :users, through: :tests_users
```
вот [пост](https://makandracards.com/makandra/32175-don-t-forget-automatically-remove-join-records-on-has_many-through-associations) в котором на примерах показано как правильно использовать опцию `:dependent` для ассоциации `has_many through`

VidgarVii commented on app/models/test.rb:
```diff
@@ -1,4 +1,9 @@
+  has_many :questions
+  belongs_to :category
   def self.by_category(category_title)
     joins('JOIN categories ON categories.id = tests.category_id')
```
с использованием ассоциации `joins` можно сильно сократить
в задании "Ассоциации моделей" самое время это сделать

***
```ruby
  joins(:category)
    .where(categories: { title: category_title })
    .order(title: :desc)
    .pluck(:title)
```

VidgarVii commented on app/models/user.rb:
```diff
@@ -1,4 +1,7 @@
+  has_many :tests_users
+  has_many :tests, through: :tests_users
   def passed_tests(test_level)
     Test
```
здесь в методе можно значительно упростить код
у тебя теперь есть ассоциация, которая вернет список всех тестов пользователя

VidgarVii commented on app/models/user.rb:
```diff
@@ -1,4 +1,7 @@
 class User < ApplicationRecord
+  has_many :tests, inverse_of: 'author'
+  has_many :tests_users
+  has_many :tests, through: :tests_users
```
В классе две ассоциации с одинковым именем `tests`
работать будет только одна
будет правильно одну из ассоциаций переименовать (к примеру назвать одну `created_tests` или другую `passed_tests`, с первой наверное будет проще разобраться, потому что она прямая, в то время как вторая через соединительную таблицу)
посмотри в документации о том как правильно создать ассоциацию не совпдадающую по имени с ассоциируемой моделью

VidgarVii commented on app/models/user.rb:
```diff
@@ -1,4 +1,7 @@
 class User < ApplicationRecord
+  has_many :tests, inverse_of: 'author'
```
когда внеший ключ не совпадает с именем таблицы, то опцию `foreign_key:` рекомендуется использовать в обеих связываемых моделях, т.е. также, как в модели Тест обозначено `foreign_key: :author_id`, также и в модели Юзер, правильно будет в `has_many` передать эту же опцию

`inverse_of:` выполняет несколько другую работу (судя по документации) - она оптимизирует запросы к базе, с той позиции, что если объекты уже подгружены, то повторный запрос при вызове ассоциации не отрпавляется

VidgarVii commented on db/migrate/20221101194543_create_tests.rb:
```diff
@@ -3,8 +3,8 @@ def change
       t.integer :level, default: '1', null: false
-      t.references :author, null: false, foreign_key: { to_table: :users }
-      t.references :category, null: false, foreign_key: true
+      t.belongs_to :author, null: false, index: true, foreign_key: { to_table: :users }
```
нет смысла менять references на belongs_to
references  более свежий и часто используемый вариант
references  index создаёт по умолчанию

RamzanBersanov committed:
На ветке lesson5
 Изменения, которые будут включены в коммит:
	изменено:      app/models/answer.rb
	изменено:      app/models/category.rb
	изменено:      app/models/question.rb
	изменено:      app/models/test.rb
	изменено:      app/models/tests_user.rb
	изменено:      app/models/user.rb
	изменено:      db/migrate/20221101194543_create_tests.rb
	изменено:      db/migrate/20221101194616_create_questions.rb
	изменено:      db/migrate/20221101194654_create_answers.rb
	изменено:      db/migrate/20221104235022_create_tests_users.rb

VidgarVii commented on app/models/answer.rb:
```diff
@@ -1,3 +1,5 @@
 class Answer < ApplicationRecord
+
```
зачем тут эти отступы не в рельсе ни в стайл гайде руби такие отступы не используются
https://github.com/rubocop/rails-style-guide 
https://rubystyle.guide/
я уже писал об этом
https://github.com/RamzanBersanov/testguru/pull/4#discussion_r1014602228

VidgarVii commented on app/models/test.rb:
```diff
@@ -1,12 +1,15 @@
   def self.by_category(category_title)
-    joins('JOIN categories ON categories.id = tests.category_id')
-      .where(categories: {title: category_title})
+    Test.joins('JOIN categories ON categories.id = tests.category_id')
```
можно (и рекомендуется)  писать так:
```ruby
joins(...)...
```
т.е. без явного указания объекта у которого вызывается метод
потому что в данном случае объектом является класс `Test` и ему равен `self` внутри класс-метода
по дефолту, при вызове метода без указания объекта он вызывается у `self`, т.е. в текущем варианте как раз у `Test`

VidgarVii replied:
документация как использовать метод join. Смотри пример с ассоциацией
http://rusrails.ru/active-record-querying#joins

VidgarVii commented on app/models/user.rb:
```diff
@@ -1,10 +1,10 @@
-    Test
-      .joins('JOIN tests_users ON tests.id = tests_users.test_id')
-      .where(tests_users: {user_id: id}, tests: {level: test_level})
+    tests.where(level: test_level).pluck("title")
```
по ТЗ метод не должен возвращать массив названий тестов

RamzanBersanov committed:
На ветке lesson5: исправлены сиды, убраны отступы, исправлен метод в модели User

RamzanBersanov committed:
На ветке lesson5: исправлены сиды

VidgarVii commented on app/models/test.rb:
```diff
@@ -1,8 +1,15 @@
+    joins(:category)
+    .where(categories: {title: category_title})
+    .order(title: :desc)
+    .pluck("title")
```
символы всё-таки смотрятся лучше и писать быстрее. Еще они потребляют памяти меньше

VidgarVii commented on app/models/test.rb:
```diff
@@ -1,8 +1,15 @@
-      .order(title: :desc)
-      .pluck(:title)
+    joins(:category)
+    .where(categories: {title: category_title})
```
зачем надо было убирать отступы?
Все 4 строчки кода это 1 вызов, который можно было написать в 1 длинную строчку. Чтобы код помещался на страницу и читался легко, каждый следующий вызов метода переносят на новую строчку с отступом в 2 пробела.
Чтобы было понятно какой код к чему относится 
Всегда используй рубокоп!

RamzanBersanov committed:
Update test.rb
